### PR TITLE
fix(auth): invalid parameter handling in saml strategy

### PR DESCRIPTION
**Pull request by IzumiSy**
# Background

Auth package does not work with SAML strategy as expected 

# Changes

This pull request includes significant changes to the authentication strategies in the `packages/auth` directory. The main changes involve the refactoring of the SSO strategy into separate OIDC and SAML strategies, modifications to the `AbstractStrategy` type, and updates to the `callback` methods in various strategies. 

Refactoring of SSO Strategy:
* [`packages/auth/src/core/strategies/sso.ts`](diffhunk://#diff-e2c6d19931678e76efce305cd3b25871f85169c7f9b4c4dd77ec6d13cf3ec9caL1-L50): The SSO strategy has been refactored into separate OIDC and SAML strategies. The `ssoStrategyBuilder` function and the `OIDCStrategy` and `SAMLStrategy` classes have been removed.
* [`packages/auth/src/core/strategies/oidc.ts`](diffhunk://#diff-ac2c2eb8f1ad1d0983cd91703f3ad89f7e0ffcfa42be06a3595c3f79bbd01acaR1-R45): Added a new file for the OIDC strategy, implementing the `AbstractStrategy` interface.
* [`packages/auth/src/core/strategies/saml.ts`](diffhunk://#diff-b8d36342ffe446a273c9df5032f5ccf6e95d6b48f7af6629ac54e309eb374c14R1-R48): Added a new file for the SAML strategy, implementing the `AbstractStrategy` interface.
* [`packages/auth/src/core/strategies/default.ts`](diffhunk://#diff-d87c6ddb37c5b101546df341c04effdf48ec08ef743ae0bd889efb6817a87541L1-R1): Updated the import statement for `OIDCStrategy` due to the refactoring.

Modifications to `AbstractStrategy` Type:
* [`packages/auth/src/core/strategies/abstract.ts`](diffhunk://#diff-fdd28aa80995655ae8d74435ce10f1bbd8304cd903706bebd5a85cdf59f53dd5L3-R3): The `CallbackResult` type has been exported and the `callback` method in the `AbstractStrategy` type has been modified to accept a `Request` object and return a `Promise<CallbackResult>`. [[1]](diffhunk://#diff-fdd28aa80995655ae8d74435ce10f1bbd8304cd903706bebd5a85cdf59f53dd5L3-R3) [[2]](diffhunk://#diff-fdd28aa80995655ae8d74435ce10f1bbd8304cd903706bebd5a85cdf59f53dd5L113-R116)

Updates to `callback` Methods:
* [`packages/auth/src/core/strategies/minitailor.ts`](diffhunk://#diff-0930b04154634e876e843a0b9e37547ae1e62013aff478bc0343ad4608857a80L35-R36): The `callback` method in the `MinitailorStrategy` class has been updated to accept a `Request` object.
* [`packages/auth/src/server/middleware/callback.ts`](diffhunk://#diff-fdab66646daaaff4f963beac1e6762442f0527b24aa57617ebbace8154afcb84L14-R14): The `callbackHandler` function has been updated to pass a `Request` object to the `callback` method and to handle the returned promise. The HTTP status code for the redirect response has been changed to 301. [[1]](diffhunk://#diff-fdab66646daaaff4f963beac1e6762442f0527b24aa57617ebbace8154afcb84L14-R14) [[2]](diffhunk://#diff-fdab66646daaaff4f963beac1e6762442f0527b24aa57617ebbace8154afcb84L34-R31)

Other Changes:
* [`packages/auth/README.md`](diffhunk://#diff-267aa09f91d442d8d9a4868e2fd7948555387a236edace3b0c72b35948da90a8R64): Added a comment to the `login` function call to specify the strategy name.
* [`packages/auth/src/core/config.ts`](diffhunk://#diff-5166aeb24926154c34a748222be317cf042e0b60df932bc3124bf2a859196653L1-R2): Updated the import statements for `OIDCStrategy` and `SAMLStrategy` due to the refactoring.
* [`packages/auth/src/core/strategies/strategy.test.ts`](diffhunk://#diff-0b1d3f52b5b7b549e7e93c96f1c99ae7abbbc9217191b5f5c4ab565c84970e7cR1-R6): Updated the tests to accommodate the changes in the strategies and the `callback` methods. [[1]](diffhunk://#diff-0b1d3f52b5b7b549e7e93c96f1c99ae7abbbc9217191b5f5c4ab565c84970e7cR1-R6) [[2]](diffhunk://#diff-0b1d3f52b5b7b549e7e93c96f1c99ae7abbbc9217191b5f5c4ab565c84970e7cL20-R37)

### CODE REVIEW

File: packages/auth/src/core/strategies/strategy.test.ts [hunk 1]
```diff
@@ -17,17 +20,20 @@ describe("redirection", () => {
...
     },
   } as const;
 
-  it.each(Object.keys(cases))("initializes authentication (%s)", (k) => {
+  it.each(Object.keys(cases))("initializes authentication (%s)", async (k) => {
     const key = k as keyof typeof cases;
-    const strategy = cases[key].builder();
-    const result = strategy.authenticate(mockAuthConfig, {
+    const strategy = cases[key].builder() as AbstractStrategy;
+    const result = await strategy.authenticate(mockAuthConfig, {
```
Comment (IzumiSy) [inline packages/auth/src/core/strategies/strategy.test.ts:30]: ここのasとawaitはコードを見ると不要に見えました。

```suggestion
    const strategy = cases[key].builder();
    const result = strategy.authenticate(mockAuthConfig, {
```

File: packages/auth/src/server/middleware/callback.ts [hunk 2]
```diff
@@ -31,7 +28,7 @@ export const callbackHandler: RouteHandler = async ({
       userID: session.user_id,
     }));
 
-  const redirection = NextResponse.redirect(config.appUrl(redirectUri));
+  const redirection = NextResponse.redirect(config.appUrl(redirectUri), 301);
```
Comment (IzumiSy) [inline packages/auth/src/server/middleware/callback.ts:31]: コメントが欲しいですので、以下でお願いします。

```suggestion
  // Next.js server redirection uses 307 and 308 to preserve the request method.
  // However the auth package always has to redirect users to the destination pages, so here intentionally uses 301 to make it GET request.
  const redirection = NextResponse.redirect(config.appUrl(redirectUri), 301);
```

File: packages/auth/src/core/strategies/oidc.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import { NextRequest } from "next/server";
```
Comment (IzumiSy) [inline packages/auth/src/core/strategies/oidc.ts:1]: SAML同様に `next/server` に依存しないように修正お願いします。

**Review by Mistat — APPROVED** [review]
LGTM

File: packages/auth/src/core/strategies/oidc.test.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import { expect } from "vitest";
+import { OIDCStrategy } from "@core/strategies/oidc";
+import { Config } from "@core";
+
+describe("oidv strategy", () => {
```
Comment (IzumiSy) [inline packages/auth/src/core/strategies/oidc.test.ts:5]:
```suggestion
describe("oidc strategy", () => {
```

File: packages/auth/src/core/strategies/oidc.test.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    const res = strategy.callback(config, req);
+    expect(res.payload.get("code")).toBe("code");
+    expect(res.payload.get("redirect_uri")).toBe(
+      "http://localhost:3000/__auth/callback/oidc",
+    );
+  });
+  it("oidc callback error", () => {
+    expect(() =>
+      strategy.callback(config, new Request("http://localhost:3000")),
+    ).toThrow("code and redirectURI should be filled");
```
Comment (IzumiSy) [inline packages/auth/src/core/strategies/oidc.test.ts:31]: ここは既存のエラー型でアサーションして欲しいので `toThrowError` でお願いします。

```suggestion
    ).toThrowError(paramsError());
```

File: packages/auth/src/core/strategies/minitailor.test.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import { afterEach } from "vitest";
+import { MinitailorStrategy } from "@core/strategies/minitailor";
+import { Config } from "@core";
+
+describe("minitailor strategy", () => {
+  beforeEach(() => {
+    vi.spyOn(global, "fetch").mockImplementation(async () =>
+      Promise.resolve(
+        new Response('{ "id_token": "test", "redirect_path": "/" }', {
+          status: 500,
```
Comment (IzumiSy) [inline packages/auth/src/core/strategies/minitailor.test.ts:10]: ここって500エラーでいいんですかね?

**Pull request merged**